recreate existing scanning ordering without recursion

benjamin-confino opened the pull request:
Fixes #28616

This creates a new algorithm that simulates the recursive depth first scanning algorithm we previously used with a visitor. The behaviour is not entirely identical. Previously if we had a loop A -> B -> A. We would scan B,A, now we will scan A,B. Given that we were already within a loop with an order influenced by whatever comes out of sets first I think its better to allow that change than to introduce what I suspect would be a messy piece of code to recreate the old system.

benjamin-confino commented:
Build started after the latest push (which was changing a class name)

Azquelt commented on dev/com.ibm.ws.org.apache.commons.collections4/bnd.bnd:
```diff
@@ -0,0 +1,15 @@
+#*******************************************************************************
+-include= jar:${fileuri;${repo;org.apache.commons:commons-collections4;4.4;EXACT}}!/META-INF/MANIFEST.MF,bnd.overrides
+
+-buildpath: org.apache.commons:commons-collections4;version=4.4
```
This is a 2.5MB jar, we probably can't justify adding it as a dependency. Also, when adding a bundle you also need to reference it in the feature files.

It should be easy enough to duplicate the functionality of `PeekingIterator`.

benjamin-confino replied:
We actually ship apache collections (not collections4) already. I was planning to see about moving us to the later version in a follow-up.

Azquelt commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveVisitor.java:
```diff
@@ -0,0 +1,71 @@
+                if (!bda.hasBeenVisited()) {
+                    Iterator<WebSphereBeanDeploymentArchive> children = bda.visit();
+                    addNextStack(children);
+                } else {
+                    if (bda.hasBeenScanned()) {
+                        bda.scan();
+                    }
+                    iterator.next();
```
I don't think this part works for two reasons:
1. it should be `if (!bda.hasBeenScanned()) {`
2. the `FilterIterator` will filter out the BDA because it's now been visited

benjamin-confino replied:
Regarding two, I checked the code of PeekingIterator before coding this, and the way it works is that when you call peek it calls the decorated iterator and caches the result. So we still have access to the BDA when we need it.

Azquelt replied:
Ah! I see, yes, it would match the filter when you call `peek()` the first time, just not the second time.

benjamin-confino committed:
use a visitor for scanning bdas

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

benjamin-confino commented:
The last change was comments and such, but just to be absolutely sure I didn't make a mistake:

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

Azquelt commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveVisitor.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * LOOP START
+     * Look at the current element from the iterator at the top of the stack, marking it visited:
+     * If it has valid (not visited or runtime extension) children place an iterator of its children on the top of its stack. GOTO LOOP START
```
Why do runtime extensions need to be treated specially?

Azquelt replied:
I see one point where the old code was skipping runtime extensions, and I can see where you've added in "now scan the runtime extensions", but I can't see where runtime extensions got scanned in the old code.

Azquelt replied:
Looked at this again - I think the old logic was "don't scan children of runtime extensions", but I think the extension itself would still be scanned when it was encountered as a child of another archive?

In the new code, I think that runtime extensions won't be scanned until the end.

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

Azquelt committed:
Attempt to simplify the recursive scanning of BDAs

I think this algorithm ends up being equivalent, but I hope it's easier
to read and reason about.

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

benjamin-confino commented:
#libby

tevans78 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java:
```diff
@@ -1132,6 +1118,34 @@ public ClassLoader run() {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+                Tr.exit(tc, getHumanReadableName() + " visited, found these children: " + accessibleBDAs);
+            }
+            return accessibleBDAs.iterator();
```
a) I don't really like methods with multiple return statements
b) You've got manual exit trace before the return but then the return statement makes another method call

benjamin-confino committed:
unit test

benjamin-confino commented:
#build
#spawn.fullfat.buckets=io.openliberty.cdi.3.0.internal_fat, io.openliberty.cdi.4.0.internal_fat, io.openliberty.cdi.4.1.internal_fat, io.openliberty.cdi.core_fat, com.ibm.ws.cdi.2.0_fat, com.ibm.ws.cdi.annotations_fat, com.ibm.ws.cdi.api_fat, com.ibm.ws.cdi.beansxml_fat, com.ibm.ws.cdi.beansxml.implicit_fat, com.ibm.ws.cdi.ejb_fat, com.ibm.ws.cdi.extension_fat, com.ibm.ws.cdi.jee_fat, com.ibm.ws.cdi.jndi_fat, com.ibm.ws.cdi.lifecycle_fat, com.ibm.ws.cdi.noncontextual_fat, com.ibm.ws.cdi.third.party_fat, com.ibm.ws.cdi.visibility_fat

Azquelt commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/AbstractBeanDeploymentArchive.java:
```diff
@@ -0,0 +1,78 @@
+import com.ibm.ws.cdi.internal.interfaces.CDIArchive;
+import com.ibm.ws.cdi.internal.interfaces.WebSphereBeanDeploymentArchive;
+
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ * IBM Corporation - initial API and implementation
+ *******************************************************************************/
+//this class exists to be extended by unit tests
+public abstract class AbstractBeanDeploymentArchive implements WebSphereBeanDeploymentArchive {
```
This copyright header is in the wrong place.

Azquelt replied:
I don't like adding extra abstract classes just for unit testing, it adds extra complexity and makes it harder to find the relevant piece of code when debugging.

We should restructure the algorithm so that it's possible to unit test it without doing this.

Azquelt commented on dev/com.ibm.ws.cdi.weld/test/com/ibm/ws/cdi/impl/test/BDAScanningTest.java:
```diff
@@ -0,0 +1,125 @@
+        //create a bunch of mock libs
+        Set<MockBeanDeploymentArchive> libs = new HashSet<MockBeanDeploymentArchive>();
+        for (int i = 0; i < 10; i++) {
+            libs.add(new MockBeanDeploymentArchive(ArchiveType.SHARED_LIB, 4, 13, "" + id.getAndIncrement()));
```
The meaning of these numbers is not clear in context.

Azquelt commented on dev/com.ibm.ws.cdi.weld/test/com/ibm/ws/cdi/impl/test/BDAScanningTest.java:
```diff
@@ -0,0 +1,125 @@
+        MockWebSphereCDIDeployment deployment = new MockWebSphereCDIDeployment(new ArrayList<WebSphereBeanDeploymentArchive>(allBDAs));
+        deployment.scan();
+
+        allBDAs.stream().forEach(bda -> bda.assertScanned());
+        allBDAs.stream().forEach(bda -> bda.seenInAcceptableOrder());
```
It's not easy to see what these checks are actually checking and how the acceptable order is defined.

In general, it should be easy to read the test and understand what it's asserting.

Azquelt commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/AbstractBeanDeploymentArchive.java:
```diff
@@ -0,0 +1,78 @@
+
+    protected abstract String getHumanReadableName();
+
+    protected abstract Set<WebSphereBeanDeploymentArchive> getDescendetBDAs();
```
Misspelling

Azquelt commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/AbstractBeanDeploymentArchive.java:
```diff
@@ -0,0 +1,78 @@
+
+            toReturn = accessibleBDAs.iterator();
+        } else {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+                Tr.exit(tc, getHumanReadableName() + " visited, found these children: []");
+            }
+            toReturn = Collections.emptyListIterator();
+        }
+
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+            String children = getArchive().getType() != ArchiveType.RUNTIME_EXTENSION ? accessibleBDAs.toString() : "[]";
+            Tr.exit(tc, getHumanReadableName() + " visited, found these children: " + children);
+        }
```
You've got two `exit` traces

benjamin-confino committed:
code review suggestions batch one

benjamin-confino committed:
code review comments batch two

tevans78 commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/WebSphereBeanDeploymentArchive.java:
```diff
@@ -4,7 +4,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
```
Copyright date

tevans78 commented on dev/com.ibm.ws.cdi.weld/test/com/ibm/ws/cdi/impl/test/BDAScanningTest.java:
```diff
@@ -0,0 +1,167 @@
+package com.ibm.ws.cdi.impl.test;
```
Missing header

benjamin-confino commented:
Closing this for now becuase it turns out fixing this only causes a similar exception to come out of weld. I'll reopen this if weld ever change their code. 

For now the solution is just to configure a larger maximum stack
